File: go/src/otest/report.go
package otest

import (
	"encoding/csv"
	"fmt"
	"os"
	"strings"
)

const reportsDir = "gen/data"

func RemoveReports(name string) {
	entries, err := os.ReadDir(reportsDir)
	if err != nil {
		panic(err)
	}

	for _, file := range entries {
		if strings.HasPrefix(file.Name(), name) {
			path := fmt.Sprintf("%s/%s", reportsDir, file.Name())
			if err := os.Remove(path); err != nil {
				panic(err)
			}
		}
	}
}

func WriteReport(name string, data []Measurement) {
	fileName := fmt.Sprintf("%s/%s.csv", reportsDir, name)
	f, err := os.OpenFile(fileName, os.O_TRUNC|os.O_CREATE|os.O_WRONLY, 0666)
	if err != nil {
		panic(err)
	}
	defer f.Close()

	w := csv.NewWriter(f)
	defer w.Flush()

	w.Write([]string{"n", "time"})

	for _, m := range data {
		w.Write([]string{
			fmt.Sprint(m.N),
			fmt.Sprint(m.Time.Nanoseconds()),
		})
	}
}

File: rust/src/otest/report.rs
use std::error::Error;

use super::runner::Measurement;

const REPORTS_DIR: &str = "./gen/data";

pub fn remove_reports(name: &str) -> Result<(), Box<dyn Error>> {
    for entry in std::fs::read_dir(REPORTS_DIR).unwrap() {
        if entry
            .as_ref()
            .unwrap()
            .file_name()
            .to_str()
            .unwrap()
            .starts_with(name)
        {
            std::fs::remove_file(entry.unwrap().path())?;
        }
    }

    return Ok(());
}

pub fn write_report(name: &str, data: Vec<Measurement>) -> Result<(), Box<dyn Error>> {
    let file_name = format!("./gen/data/{}.csv", name);
    let mut wtr = csv::Writer::from_path(file_name)?;

    wtr.write_record(["n", "time"])?;
    for d in data {
        wtr.write_record([d.n.to_string(), d.time.as_nanos().to_string()])?;
    }

    wtr.flush()?;

    return Ok(());
}
